examples/tagging/main_train_tagging.py

```diff
@@ -0,0 +1,215 @@
+
+    model: BiRecurrentConvCRF = \
+        BiRecurrentConvCRF(word_vocab=text_extractor.get_dict(),
+                           char_vocab_size=char_extractor.size(),
+                           tag_vocab_size=output_extractor.size(),
+                           config_model=config_model)
+    model.to(device)
+
+    optim: Optimizer = SGD(model.parameters(),
+                           lr=config_model.learning_rate,
+                           momentum=config_model.momentum,
+                           nesterov=True)
+
+    predictor = Predictor(
+        batch_size=train_preprocessor.config.dataset.batch_size,
+        model=model,
+        predict_forward_fn=predict_forward_fn,
+        feature_resource=train_preprocessor.request,
+        cross_pack=False)
+    evaluator = CoNLLNEREvaluator()
+
+    val_reader = CoNLL03Reader(cache_in_memory=True)
+    val_pl: Pipeline = Pipeline()
+    val_pl.set_reader(val_reader)
+    val_pl.add(predictor)
+    val_pl.add(evaluator)
+
+    epoch = 0
+    train_err: int = 0
+    train_total: float = 0.0
+    train_sentence_len_sum: float = 0.0
+    val_scores: List = []
+    output_file = "tmp_eval.txt"
+    score_file = "tmp_eval.score"
+    scores: Dict[str, float] = {}
+
+    logger.info("Start training.")
+    while epoch < config_data.num_epochs:
+        epoch += 1
+
+        # Get iterator of preprocessed batch of train data
+        train_batch_iter: Iterator[Batch] = \
+            train_preprocessor.get_train_batch_iterator()
+
+        for batch in tqdm(train_batch_iter):
+            word = batch["text_tag"]["data"]
+            char = batch["char_tag"]["data"]
+            output = batch["output_tag"]["data"]
+            word_masks = batch["text_tag"]["masks"][0]
+
+            optim.zero_grad()
+
+            loss = model(word, char, output, mask=word_masks)
+
+            loss.backward()
+            optim.step()
+
+            batch_train_err = loss.item() * batch.batch_size
+
+            train_err += batch_train_err
+            train_total += batch.batch_size
+            train_sentence_len_sum += \
+                torch.sum(batch["text_tag"]["masks"][0]).item()
+
+        logger.info("%dth Epoch training, "
+                    "total number of examples: %d, "
+                    "Average sentence length: %0.3f, "
+                    "loss: %0.3f",
+                    epoch, train_total,
+                    train_sentence_len_sum / train_total,
+                    train_err / train_total)
+
+        train_err = 0
+        train_total = 0.0
+        train_sentence_len_sum = 0.0
+
+        val_pl.run(config_data.val_path)
+
+        logger.info("%dth Epoch evaluating, "
+                    "val result: %s",
+                    epoch, evaluator.get_result())
+
+        # Save training state to disk
+        # train_preprocessor.save_state(config_data.train_state_path)
+        # torch.save(model, config_model.model_path)
```

Comment:
Un comment this?

Reply:
OK

Reply:
Those will be replaced with new apis (i.e. each extractor/converter saves/loads its own state)